assets/js/pixels.js

```diff
@@ -3,42 +3,206 @@
+            seenList[concat] = true;
+        }
+
 
         // We use the Beacon API if it's available for more accurate logging.
         // Reference: https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API
         if ('sendBeacon' in navigator) {
-            navigator.sendBeacon(pixelPath + data);
+            navigator.sendBeacon(pixelPath);
         } else {
             // Otherwise we go with creating an <img> element
             // and setting the src attribute with the GIF.
             let pixel = $('<img>');
-            pixel.attr('src', pixelPath + data);
+            pixel.attr('src', pixelPath);
         }
     }
+
+    // Helper: requestTick
+    // Better performance on scrolling by using debouncing.
+    // Reference: https://gomakethings.com/debouncing-events-with-requestanimationframe-for-better-performance/
+    function requestTick() {
+        if (!ticking) {
+            requestAnimationFrame(updateScrollInfo);
+        }
+        ticking = true;
+    }
+
+    // Helper: updateScrollInfo
+    // ---------------
+    // Check how far down the user has scrolled.
+    // Example: blog.half.quora.google-filter-bubble-study
+    let imageUrlCache = {};
+    function updateScrollInfo() {
+        let progressMax = lastDocumentHeight - lastWindowHeight;
+        let percentage = lastScrollY / progressMax;
+
+        articleImagesEl.each(function() {
+            if(!imageUrlCache[this.src]) {
+                imageUrlCache[this.src] = this.src.split('/').pop();
+            }
+
+            if (elementIsVisibleInViewport(this, true)) {
+                firePixel('image', source, pathname,
+                          sanitizeUrl(imageUrlCache[this.src]),
+                          {once: true});
+            }
+        });
+
+        ticking = false;
+    }
+
+    // Helper: sanitizeUrl
+    // Sanitzes a given URL by removing unsafe characters.
+    // E.g. 'https://reddit.com/r/duckduckgo' -> 'https-reddit-com-r-duckduckgo'
+    function sanitizeUrl(url) {
+        return url
+        // strip leading/trailing slash
+            .replace(/^\/|\/$/, '')
+        // strip unsafe chars for grafana
+            .replace(/[^a-z0-9_-]+/ig, '-')
+        // strip underscores as well
+            .replace(/_/g,'-');
+    }
+
+    // Helper: elementIsVisibleInViewport
+    // Source: https://github.com/30-seconds/30-seconds-of-code#elementisvisibleinviewport-
+    // Figures out if an element (in our case images) are in view so that we
+    // can fire the pixels.
+    function elementIsVisibleInViewport(el, partiallyVisible) {
+        partiallyVisible = partiallyVisible || false;
+
+        const elBounds = el.getBoundingClientRect();
+        const top = elBounds.top;
+        const left = elBounds.left;
+        const bottom = elBounds.bottom;
+        const right = elBounds.right;
+
+        const innerHeight = window.innerHeight;
+        const innerWidth = window.innerWidth;
+
+        return partiallyVisible
```

Comment:
Might be best to avoid the ternary here for the sake of legibility

```js
if (partiallyVisible) {
    return ( (...) || (...) ) && ( (...) || (...) );
} else {
    return ...;
}
```